lib/IRGen/IRGenFunction.cpp

```diff
@@ -513,3 +513,249 @@ llvm::Value *IRGenFunction::alignUpToMaximumAlignment(llvm::Type *sizeTy, llvm::
+  atomicStore->setAtomic(llvm::AtomicOrdering::Release,
+                         llvm::SyncScope::System);
+  out.add(unsafeContinuation);
+}
+
+void IRGenFunction::emitAwaitAsyncContinuation(
+    SILType unsafeContinuationTy, bool isIndirectResult,
+    Explosion &outDirectResult, llvm::BasicBlock *&normalBB,
+    llvm::PHINode *&optionalErrorResult, llvm::BasicBlock *&optionalErrorBB) {
+  assert(AsyncCoroutineCurrentContinuationContext && "no active continuation");
+  auto pointerAlignment = IGM.getPointerAlignment();
+
+  // First check whether the await reached this point first. Meaning we still
+  // have to wait for the continuation result. If the await reaches first we
+  // abort the control flow here (resuming the continuation will execute the
+  // remaining control flow).
+  auto contAwaitSyncAddr =
+      Builder.CreateStructGEP(AsyncCoroutineCurrentContinuationContext, 1);
+  auto null = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 0);
+  auto one = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 1);
+  auto results = Builder.CreateAtomicCmpXchg(
+      contAwaitSyncAddr, null, one,
+      llvm::AtomicOrdering::AcquireRelease /*success ordering*/,
+      llvm::AtomicOrdering::Monotonic /* failure ordering */,
+      llvm::SyncScope::System);
+  auto firstAtAwait = Builder.CreateExtractValue(results, 1);
+  auto contBB = createBasicBlock("await.async.maybe.resume");
+  auto abortBB = createBasicBlock("await.async.abort");
+  Builder.CreateCondBr(firstAtAwait, abortBB, contBB);
+  Builder.emitBlock(abortBB);
+  {
+    // We are first to the sync point. Abort. The continuation's result is not
+    // available yet.
+    emitCoroutineOrAsyncExit();
+  }
+
+  auto contBB2 = createBasicBlock("await.async.resume");
+  Builder.emitBlock(contBB);
+  {
+    // Setup the suspend point.
+    SmallVector<llvm::Value *, 8> arguments;
+    arguments.push_back(AsyncCoroutineCurrentResume);
+    auto resumeProjFn = getOrCreateResumePrjFn();
+    arguments.push_back(
+        Builder.CreateBitOrPointerCast(resumeProjFn, IGM.Int8PtrTy));
+    arguments.push_back(Builder.CreateBitOrPointerCast(
+        getOrCreateAwaitAsyncSupendFn(), IGM.Int8PtrTy));
+    arguments.push_back(AsyncCoroutineCurrentResume);
+    arguments.push_back(
+        Builder.CreateBitOrPointerCast(getAsyncTask(), IGM.Int8PtrTy));
+    arguments.push_back(
+        Builder.CreateBitOrPointerCast(getAsyncExecutor(), IGM.Int8PtrTy));
+    arguments.push_back(Builder.CreateBitOrPointerCast(
+        AsyncCoroutineCurrentContinuationContext, IGM.Int8PtrTy));
+    auto *id = Builder.CreateIntrinsicCall(llvm::Intrinsic::coro_suspend_async,
+
+                                           arguments);
+    auto results = Builder.CreateAtomicCmpXchg(
+        contAwaitSyncAddr, null, one,
+        llvm::AtomicOrdering::AcquireRelease /*success ordering*/,
```

Comment:
The ordering needs to be swapped, we continue with control flow if the swap failed (i.e all information is available: both the await point was reached and the continuation was resumed).